Consider PAR1/2 on Chr X as autosomal

rollf opened the pull request:
This PR addresses issue [#789](https://github.com/etal/cnvkit/issues/784). It adds the optional feature to treat the PAR1/2 on chromosome X as autosomal, i.e. when requesting autosomal regions these regions are included and on the other hand, when requesting "chromosome X", they are excluded. For chr Y, the PAR1/2 regions are excluded when requesting "chromosome Y" since they cannot be covered if all reads map to PAR1/2 on X. The results have already been shown in the issue mentioned above.


This PR contains the following changes:
* Add new command line argument `--diploid-parx-genome` which can be set to `grch37` or `grch38` in order to treat PAR1/2 GRCh37/38 regions on chromosome X as autosomal.
* This argument is added to all subcommands where necessary and forwarded.
* The method `GenomicArray.autosomes()` has been re-implemented in `CopyNumArray.autosomes()` to be able to transparently include/exclude PAR1/2 on demand.
* Conversely, wherever chromosome X regions have been requested (ex: `(cnarr.chromosome == cnarr._chr_x_label)`), this has been replaced by a call to `CopyNumArray.chr_x_filter()` which includes/excludes PAR1/2 as needed. Same for chromosome Y.
* No differentiation is done based on the sex of the sample.
* This approach has the advantage of having only one place where if/else semantics have to implemented. Whenever the code treats chromosome X, one doesn't have to reason about whether PAR1/2 is include or not.
*  This approach has disadvantages: If enabled, the regions for chrX will be split even for female samples (where it doesn't make sense, as opposed to for male samples [see #784 for an explanation]). Also, one has to rely on the user to write `chrx = self[self.chrx_filter()]` instead of `self[self.chromosome == self.chr_x_label]` (or even `self[self.chromosome == "chrX"]) - but it's in the nature of Python to give full access to object members. 
* Fix a bug in `call.absolute_dataframe()`.
* Add `read_ga` similar to `read_cna()` and use where appropriate (=GenomicArray tests)
* I fixed the version of pomegranate to 0.14.8. See comments.

I will add some further review comments.

rollf commented on cnvlib/call.py:
```diff
@@ -148,15 +148,9 @@ def absolute_threshold(cnarr, ploidy, thresholds, is_reference_male):
-        absolutes[i] = _log2_ratio_to_absolute(
-            row.log2, ref_copies, expect_copies, purity)
-    return absolutes
+    df = absolute_dataframe(cnarr, ploidy, purity, is_reference_male, is_sample_female)
```
`absolute_clonal()` uses a subset of the results from `absolute_dataframe()` so I figured this could be refactored here.

rollf commented on cnvlib/call.py:
```diff
@@ -170,18 +164,9 @@ def absolute_pure(cnarr, ploidy, is_reference_male):
 def absolute_dataframe(cnarr, ploidy, purity, is_reference_male, is_sample_female):
     """Absolute, expected and reference copy number in a DataFrame."""
-    absolutes = np.zeros(len(cnarr), dtype=np.float_)
-    reference_copies = expect_copies = np.zeros(len(cnarr), dtype=np.int_)
```
This is a bug here (at least, I don't think this behavior was intentional). Consider this:

![absolute_dataframe_bug](https://user-images.githubusercontent.com/58295931/220121838-6b12b021-e621-422a-905d-ca893f8645e1.png)

(I.e. any change of either `reference_copies` or `expect_copies` will get reflected in both.)

rollf commented on cnvlib/call.py:
```diff
@@ -231,16 +216,18 @@ def _reference_expect_copies(chrom, ploidy, is_sample_female, is_reference_male)
-    else:
-        ref_copies = exp_copies = ploidy
-    return ref_copies, exp_copies
+    df = cnarr.copy().data
```
Direct access of the `data` member. I initially had implemented a `CopyNumArray.get_copied_dataframe()` but realized that at several places throughout the code base `data` is accessed directly. One could replace all those with a dedicated method, though.

rollf commented on skgenome/gary.py:
```diff
@@ -205,17 +205,21 @@ def sample_id(self):
         if not is_auto.any():
             # The autosomes, if any, are not named with plain integers
             return self
-        if also:
```
This condition doesn't really play nice with pandas. Thus I changed the default value from `()` to `None` and made the check explicit.

rollf commented on test/test_commands.py:
```diff
@@ -207,6 +207,77 @@ def test_chrom_means(segments):
+        return ret
+
+    def test_call_absolute_clonal_and_dataframe(self):
+        """Test some helper methods for `call`."""
```
I added this test here before fix/refactoring `call.absolute_clonal()`/`absolute_dataframe()` so that I'm sure I'm improving things here.

etal replied:
Excellent, thanks

rollf commented on test/test_genome.py:
```diff
@@ -11,15 +11,15 @@
 
     def setUp(self):
-        self.ex_cnr = read('formats/reference-tr.cnn')
+        self.ex_cnr = read_ga('formats/reference-tr.cnn')
```
This test is for `GenomicArray` so we should not test a `CopyNumArray` object.

rollf commented:
@etal Could you have a look into this when time permits?

(I believe the conflicts arose due to new formatting)

rollf commented on test/test_genome.py:
```diff
@@ -12,7 +12,7 @@
 import pandas as pd
 
-from cnvlib import read
+from cnvlib import read_ga as read
```
Make sure to actually test `GenomicArray` not `CopyNumArray`

etal replied:
How about importing `read_ga` with is original name and updating the function calls within the unit tests to `read_ga`?

etal commented:
Thanks! I'm reviewing this PR now. Sorry for forcing a rebase, but hopefully the codebase is a little easier to work with now.

etal commented on cnvlib/commands.py:
```diff
@@ -86,6 +86,12 @@ def public(fn):
 
+# Shared parameters
+def add_argument_treat_par_on_chrx_as_autosomal_for_genome_build(P):
+    P.add_argument('--treat-par-on-chrx-as-autosomal-for-genome-build',
```
Should `P` be `parser`? I see an urge to refactor commands.py more substantially this way and collect other reused options into functions like this, but it can be left for the future.

I see help is "todo" here -- how about: "Treat the PAR region of human chromome X as autosomal using its known reference genome coordinates."

etal commented on cnvlib/export.py:
```diff
@@ -27,12 +27,12 @@ def label_with_gene(cnarr):
     if not filenames:
         return []
-    first_cnarr = read_cna(filenames[0])
+    first_cnarr = read_cna(filenames[0], None)
```
It's not obvious what the second argument means at this point; could it be specified as a keyword argument in each case here and below?

rollf replied:
Yes, I can do this. Depends on whether I make the param optional in the first place (your call).

etal commented on cnvlib/metrics.py:
```diff
@@ -6,7 +6,7 @@
 
 
-def do_metrics(cnarrs, segments=None, skip_low=False):
+def do_metrics(cnarrs, segments=None, skip_low=False): # todo: make this work in parallel?
```
Meaning, run metrics on each cnarr in parallel?

rollf replied:
Yes, same as for `reference`. Both commands seem ripe for parallelization.

etal commented on cnvlib/reference.py:
```diff
@@ -44,14 +45,15 @@ def bed2probes(bed_fname):
     table["log2"] = 0.0
     table["spread"] = 0.0
-    return CNA(table, {"sample_id": core.fbase(bed_fname)})
+    return CNA(table, {"sample_id": core.fbase(bed_fname)}) # todo: find all invocations of CNA() and forward param
```
What does this mean? I don't see a new parameter used here.

rollf replied:
Maybe the wording here is unfortunate. I wanted to make sure a decision is taken, not only for `read_cna` but also when invoking  the constructor (see my other comment).

etal commented on test/test_cnvlib.py:
```diff
@@ -16,24 +16,42 @@
     def test_empty(self):
         """Instantiate from an empty file."""
-        cnarr = cnvlib.read("formats/empty")
+        cnarr = cnvlib.read("formats/empty", None)
```
Could the second argument be handled by a default value of None in the signature of `cnvlib.read` instead?

rollf commented on cnvlib/cmdutil.py:
```diff
@@ -5,11 +5,20 @@
-
-def read_cna(infile, sample_id=None, meta=None):
+# todo: check all invocations of read/read_cna and make sure the parameter is forwarded.
+def read_cna(infile, treat_par_on_chrx_as_autosomal_for_genome_build, into=CNA, sample_id=None, meta=None):
```
I feel like the decision on PAR is important and I wanted to make sure one is forced to think about it when creating the `CNA`. At least here (with `read_cna()`) I could do this. When invoking the ctor directly (or even cast'ing), I cannot force the developer to take the decision on PAR.

I do understand your concern about breakage for anyone using `cnvlib` as dependency. I think in the current code base, nothing breaks for now. In any case, I'm happy to move the parameter to the end (`into`) as well and thus take a default decision. Just let me know.

rollf replied:
As for putting it into `meta`, yes, this is a good idea. This would allow to refactor the code and do the `if-else` here instead of this [line](https://github.com/etal/cnvkit/blob/0a1735753f505e5748d89e56f078cdd70177072c/cnvlib/cnary.py#L32).

etal commented:
> You haven't said whether you're globally fine with my approach but I assume it's okay then. I will change the code towards adding the `treat_par...` parameter into the `meta` dict.

Yes, I'm globally fine with your approach, and I appreciate your efforts here. My only restriction in this PR is that `read_cna(filename)` and `CNA(table)` should work as expected on a non-human genome like yeast. For human-specific behavior I prefer named keyword arguments over positionals, but you can see I haven't been consistent with that either, so I won't fuss too much about it.

rollf commented:
Hey @etal,

thanks again for your feedack. I have been working on other topics but I had also tried to make headway here. However, I was really unsatisfied with my implementation. Mainly, because I feel like I am not able ensure consistency (in terms of behavior) across the whole code base (i.e. I can't make sure that the necessary parameter(s) is/are always available when creating/dealing with `CopyNumberArray` objects). I'm currently pursuing a more global approach and will of course get back to you once I have something to show.

rollf commented:
@etal I worked through the code base again and I'm now done. Can you please carefully review this again? I have updated the initial PR comment as well as added further comments since I have changed my mind and found a better way (IMHO) to implement this optional feature. The tests are currently failing due to dependency issues (see my comment in `core.txt`). I had also used `blackd` but refrained from doing so because I felt like I'm not using the same configuration (some parts of the code were formatted which I didn't touch).

Here are 2 x 3 plots. For the male & the female sample from a mixed pool, I have plotted chr X once based on the current master branch, once based on my branch without the feature enabled and once with the feature enabled. The male bias towards a gain and the female bias towards a loss are fixed (last plot for each sex).

male:
![male01](https://github.com/etal/cnvkit/assets/58295931/6800bc65-8efd-475a-93a5-36b45cfd2671)
![male01](https://github.com/etal/cnvkit/assets/58295931/30d94140-0ba8-4f6d-b6f6-e607ddfabca0)
![male01](https://github.com/etal/cnvkit/assets/58295931/380b140b-fbec-4745-91ca-d2a0751fec69)

female:
![female01](https://github.com/etal/cnvkit/assets/58295931/e3746ce7-1ac8-4ef2-9ba1-4e7c6cf73dfe)
![female01](https://github.com/etal/cnvkit/assets/58295931/dbc952ac-e5e8-4caa-a5ed-a778b34ee569)
![female01](https://github.com/etal/cnvkit/assets/58295931/60daa5c6-47ae-4728-bbe4-925c07b77dcf)

rollf commented:
Hi @etal,

thanks for the feedback. I renamed the file as you suggested. There are still two tests failing. For me, this looks like a dependency issue (after pinning pomegranate). Let me know how you would like to handle this.

rollf commented:
Hi @etal,

I will be on a longer summer break for July & August. Do you think we could finish this before end of June? I'm happy to support you if needed but I don't see anything I could do for now.

etal commented on cnvlib/call.py:
```diff
@@ -174,54 +168,46 @@ def absolute_pure(cnarr, ploidy, is_reference_male):
 
 
-def absolute_expect(cnarr, ploidy, is_sample_female):
+def absolute_expect(cnarr, ploidy, diploid_parx_genome, is_sample_female):
```
With pandas 2's copy-on-write feature there might not be much memory overhead. Could be worth consolidating code.

rollf replied:
I realize that the `purity` parameter would be missing and beside, all the methods already invoke `get_as_dframe_and_set_reference_and_expect_copies()` internally. So I'll keep as is.

etal commented:
Sure. I'm not too fond of the tests `py3.8-min` and `py3.10-macos`. Ignoring those tests, do you feel this PR is ready to merge?

etal commented on cnvlib/call.py:
```diff
@@ -237,16 +231,22 @@ def _reference_expect_copies(chrom, ploidy, is_sample_female, is_reference_male)
+        ploidy if is_sample_female else ploidy // 2
+    )
+
+    df.loc[df.chromosome == cnarr._chr_y_label, "reference"] = ploidy // 2
```
The better fix here would be to change the parameter name `is_reference_male` to `is_haploid_x_reference` or equivalent. I changed the CLI option name from `--male-reference` to `--haploid-x-reference` in commands.py but it looks like I didn't bother renaming arguments in the deeper function signatures.

The reference profiles come in two flavors: XY (formerly "male" reference, now "haploid x") and XXY (never documented as "female"; it's just the default):
https://cnvkit.readthedocs.io/en/stable/sex.html#reference-sex-chromosome-ploidy

Why does the reference profile have Klinefelter syndrome by default? So that male samples' chrY values can be calculated, while plots still look reasonable (for female samples, copy-neutral X will look the same as the autosomes and there won't be any signal for Y). If reference chrY is 0 then we get divide-by-zero problems.

rollf replied:
Okay, I replaced occurrences of `reference_male` by `is_haploid_x_reference` (except occurrences of `args.male_reference`) and added a comment.

rollf commented:
@etal Yes, this is ready for merge IMHO.

etal commented:
All done, squashed and merged. Thanks for your contribution!

rollf commented:
Yeah! :rocket:

eesiribloom commented:
is there a reason not to implement being able to specify/force sex on input for commands like `cnvkit.py batch` ?
It seems like one of the most common issues users have with cnvkit. I myself have many female samples which are incorrectly assumed to be male.
